### DISCUSSION TITLE
memcpy seems to slower on Atari on latest build

### DISCUSSION THREAD

**Discussion by gianlucarenzi** [cc65/cc65#1945] [Q&A]
Hello everybody,
I was wondering if there are some issues on Atari part of string.h library with memcpy function.
A couple of years ago I was testing (with emulators and not real hardware) the best compromises between which is the best chunk size of buffers to be moved from internal ram space to an I/O space buffer address with C64 (using Vice X64 emulator) and Atari XL (using Atari800 emulator). Basically the code is looping transferring from 32 bytes up to the maximum chunk size (256 bytes at maximum) in one second and measuring the elapsed timing within one second. I remember perfectly C64 was transferring up to 33.8kbytes/sec running at 1 MHz, meanwhile the Atari was able to reach 52kbytes/sec having the main clock of 1.79 MHz.
Yesterday, with the latest build of cc65 and latest emulators, I had the Atari side slowing down to 40kbytes/sec. The c64 side remains the same. Can you help me to find out how to find why?

**Comment by polluks2** [comment]
Well, memcpy should not depend of these platforms at all.
```
./libsrc/geos-common/common/memcpy.s
./libsrc/pce/memcpy.s
./libsrc/common/memcpy.s

**Comment by mrdudz** [comment]
I'd make a minimal testcase, and then look at the generated code (use --add-source).

However, it sounds like this kind of code would not be different between the two platforms - and thus the platform itself is what makes the difference.

**Comment by gianlucarenzi** [comment]
Having the interleaved assembly code in a common code wrote by me I think
doesn't matter but I will do as per reference. If the issue is into the
compiler itself or into the platform code having interleaved code doesn't
matter too, because I suppose the interleaved code is only in new compiled
code and not in a compiler itself.

Maybe something in between. I am using version 4.8 of the Atari800 emulator
or Version 5.0. The issue remains between these two versions.

Today I will try my code with a real Atari to see what's happening but I am
sure enough that the issue is in the generated code by cc65 compiler...

For your information the code is compiled with -Osir flags

Il sab 10 dic 2022, 20:31 Bob Andrews ***@***.***> ha scritto:

**Comment by gianlucarenzi** [comment]
For your information I tried to compile atari800 v4.2.0 (from my github
repo), and using cc65 (from my github repo too) version V2.19 - Git
62da869e4 and SDL 1.2 (needed by atari800) on my OS X 10.14.6 Mojave on a
Core Duo System. In that case I reach the 51.959 bytes/sec using a maximum
of 223 bytes size chunk.
I will investigate further on my main Linux Debian x86_64 machine, then.
I will keep you up-to-date.

Cheers,
Gianluca

Il giorno dom 11 dic 2022 alle ore 07:00 Gianluca Renzi ***@***.***>
ha scritto:

> Having the interleaved assembly code in a common code wrote by me I think
> doesn't matter but I will do as per reference. If the issue is into the
> compiler itself or into the platform code having interleaved code doesn't
> matter too, because I suppose the interleaved code is only in new compiled
> code and not in a compiler itself.
>
> Maybe something in between. I am using version 4.8 of the Atari800
> emulator or Version 5.0. The issue remains between these two versions.
>
> Today I will try my code with a real Atari to see what's happening but I
> am sure enough that the issue is in the generated code by cc65 compiler...
>
> For your information the code is compiled with -Osir flags
>
> Il sab 10 dic 2022, 20:31 Bob Andrews ***@***.***> ha
> scritto:
>
>> I'd make a minimal testcase, and then look at the generated code (use
>> --add-source).
>>
>> However, it sounds like this kind of code would not be different between
>> the two platforms - and thus the platform itself is what makes the
>> difference.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cc65/cc65/discussions/1945#discussioncomment-4367513>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AFN6ERYJCCIMM4FEBXPZJGTWMTLC7ANCNFSM6AAAAAAS2K4CRE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

-- 
Ciao e buona giornata.

"GP! In mezzo al campo stai proprio schifoso!"
Coach M.Russo

**Comment by gianlucarenzi** [comment]
Maybe I have found what causes the issue. It was a mix between the (bad)
installation from sources and the atari800 configuration file.
After uninstalling the cc65 and re-installing it systemwide with the
repository for Debian 11 via *apt-get install *and removing the atari800
configuration file, everything is working fine now.
Anyway I must say, the latest (from github) cc65 code is not installing
properly from the sources using the way I did since 2019:
> git clone https://github.com/cc65/cc65.git
> cd cc65
cc65> make PREFIX=$HOME all
cc65> make PREFIX=$HOME install

When launching a Makefile over a project, it fails with not finding the
include directories. I must include the -I/path-to-includes (i.e. something
like $(HOME)/share/cc65/include )

For your information.

Regards,
Gianluca

Il giorno dom 11 dic 2022 alle ore 15:59 Gianluca Renzi ***@***.***>
ha scritto:

> For your information I tried to compile atari800 v4.2.0 (from my github
> repo), and using cc65 (from my github repo too) version V2.19 - Git
> 62da869e4 and SDL 1.2 (needed by atari800) on my OS X 10.14.6 Mojave on a
> Core Duo System. In that case I reach the 51.959 bytes/sec using a maximum
> of 223 bytes size chunk.
> I will investigate further on my main Linux Debian x86_64 machine, then.
> I will keep you up-to-date.
>
> Cheers,
> Gianluca
>
> Il giorno dom 11 dic 2022 alle ore 07:00 Gianluca Renzi ***@***.***>
> ha scritto:
>
>> Having the interleaved assembly code in a common code wrote by me I think
>> doesn't matter but I will do as per reference. If the issue is into the
>> compiler itself or into the platform code having interleaved code doesn't
>> matter too, because I suppose the interleaved code is only in new compiled
>> code and not in a compiler itself.
>>
>> Maybe something in between. I am using version 4.8 of the Atari800
>> emulator or Version 5.0. The issue remains between these two versions.
>>
>> Today I will try my code with a real Atari to see what's happening but I
>> am sure enough that the issue is in the generated code by cc65 compiler...
>>
>> For your information the code is compiled with -Osir flags
>>
>> Il sab 10 dic 2022, 20:31 Bob Andrews ***@***.***> ha
>> scritto:
>>
>>> I'd make a minimal testcase, and then look at the generated code (use
>>> --add-source).
>>>
>>> However, it sounds like this kind of code would not be different between
>>> the two platforms - and thus the platform itself is what makes the
>>> difference.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/cc65/cc65/discussions/1945#discussioncomment-4367513>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AFN6ERYJCCIMM4FEBXPZJGTWMTLC7ANCNFSM6AAAAAAS2K4CRE>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>
>
> --
> Ciao e buona giornata.
>
> "GP! In mezzo al campo stai proprio schifoso!"
> Coach M.Russo
>


-- 
Ciao e buona giornata.

"GP! In mezzo al campo stai proprio schifoso!"
Coach M.Russo

**Comment by mrdudz** [comment]
> Anyway I must say, the latest (from github) cc65 code is not installing properly from the sources using the way I did since 2019:

please open an issue for this problem

**Comment by gianlucarenzi** [comment]
On 12/13/22 11:49, Bob Andrews wrote:
>
>     Anyway I must say, the latest (from github) cc65 code is not
>     installing properly from the sources using the way I did since 2019:
>
> please open an issue for this problem
>
I will do ASAP
